async_grpc/server_test.cc

```diff
@@ -277,22 +264,73 @@ TEST_F(ServerTest, ProcessServerStreamingRpcTest) {
   }
   EXPECT_FALSE(client.StreamRead(&response));
   EXPECT_TRUE(client.StreamFinish().ok());
-
-  server_->Shutdown();
 }
 
 TEST_F(ServerTest, RetryWithUnrecoverableError) {
-  server_->Start();
-
   Client<GetSquareMethod> client(
       client_channel_, common::FromSeconds(5),
       CreateUnlimitedConstantDelayStrategy(common::FromSeconds(1),
                                            {::grpc::INTERNAL}));
   proto::GetSquareRequest request;
   request.set_input(-11);
   EXPECT_FALSE(client.Write(request));
+}
+
+TEST_F(ServerTest, AsyncClientUnary) {
+  std::mutex m;
+  std::condition_variable cv;
+  bool done = false;
+
+  AsyncClient<GetSquareMethod> async_client(
+      client_channel_,
+      [&done, &m, &cv](const ::grpc::Status& status,
+                       const proto::GetSquareResponse* response) {
+        EXPECT_TRUE(status.ok());
+        EXPECT_EQ(response->output(), 121);
+        {
+          std::lock_guard<std::mutex> lock(m);
+          done = true;
+        }
+        cv.notify_all();
+      });
+  proto::GetSquareRequest request;
+  request.set_input(11);
+  async_client.WriteAsync(request);
+
+  std::unique_lock<std::mutex> lock(m);
+  cv.wait(lock, [&done] { return done; });
+}
+
+TEST_F(ServerTest, AsyncClientServerStreaming) {
+  std::mutex m;
+  std::condition_variable cv;
+  bool done = false;
+  int counter = 0;
+
+  AsyncClient<GetSequenceMethod> async_client(
+      client_channel_,
+      [&done, &m, &cv, &counter](const ::grpc::Status& status,
+                                 const proto::GetSequenceResponse* response) {
+        LOG(INFO) << status.error_code() << " " << status.error_message();
+        LOG(INFO) << (response ? response->DebugString() : "(nullptr)");
+        EXPECT_TRUE(status.ok());
+
+        if (!response) {
+          {
+            std::lock_guard<std::mutex> lock(m);
+            done = true;
+          }
+          cv.notify_all();
+        } else {
+          EXPECT_EQ(response->output(), counter++);
+        }
+      });
+  proto::GetSequenceRequest request;
+  request.set_input(10);
+  async_client.WriteAsync(request);
 
```

Comment:
you could add a log line before and after the wait so a failing test would be easier to debug